File: scripts/generic_websocket_dashboard.py
"""Simple websocket client for manual communication protocol testing."""

import aiohttp
import asyncio
import json

# Example request payload for the communication protocol websocket. Adjust the
# execution_dag, target_exports, and kwargs to match the reducers you want to
# exercise.
REQUEST = {
    "docs_request": {
        "execution_dag": "writing_observer",
        "target_exports": ["docs_with_roster"],
        "kwargs": {
            "course_id": 12345678901
        },
    }
}


async def main():
    async with aiohttp.ClientSession() as session:
        async with session.ws_connect(
            "http://localhost:8888/wsapi/communication_protocol",
            timeout=5,
        ) as ws:
            await ws.send_json(REQUEST)
            async for msg in ws:
                if msg.type == aiohttp.WSMsgType.TEXT:
                    try:
                        parsed = json.loads(msg.data)
                    except json.JSONDecodeError:
                        parsed = msg.data
                    print(json.dumps(parsed, indent=2))
                elif msg.type == aiohttp.WSMsgType.ERROR:
                    print("Error received from websocket:", msg)
                    break


if __name__ == '__main__':
    asyncio.run(main())

File: scripts/generic_websocket_dashboard.js
/*
  Simple Node.js websocket client for exercising the communication protocol.

  Update the REQUEST payload to point at the execution DAG and exports you want
  to test.
*/

// Remove this line when running within a browser terminal (i.e. non-Node.js environment)
const WebSocket = require('ws');

const SERVER = 'ws://localhost:8888/wsapi/communication_protocol';

const REQUEST = {
  docs_request: {
    execution_dag: 'writing_observer',
    target_exports: ['docs_with_roster'],
    kwargs: {
      course_id: 'COURSE-123',
    }
  }
};
const socket = new WebSocket(SERVER);

socket.on('open', () => {
  console.log('Open');
  socket.send(JSON.stringify(REQUEST));
});

socket.on('message', (msg) => {
  try {
    const parsed = JSON.parse(msg.toString());
    console.log(parsed);
  } catch (error) {
    console.log(msg.toString());
  }
});

socket.on('error', (err) => {
  console.log('Error');
  console.log(err);
});

socket.on('close', (event) => {
  console.log('Close');
  if (event) {
    console.log(event);
  }
});
